File: app/src/main/java/com/mc10inc/codetest/di/ApplicationModule.java
package com.mc10inc.codetest.di;

import com.mc10inc.codetest.BuildConfig;
import com.mc10inc.codetest.CodeTestApplication;
import com.mc10inc.codetest.PreferenceManager;
import com.mc10inc.codetest.api.Mc10WebService;
import com.squareup.moshi.FromJson;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.ToJson;

import org.threeten.bp.Instant;

import java.util.TimeZone;
import java.util.UUID;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;
import timber.log.Timber;

/**
 * TODO
 *
 * @author Adam Stroud &#60;<a href="mailto:adam.stroud@gmail.com">adam.stroud@gmail.com</a>&#62;
 */
@Module
public class ApplicationModule {
    private final CodeTestApplication codeTestApplication;

    public ApplicationModule(CodeTestApplication codeTestApplication) {
        this.codeTestApplication = codeTestApplication;
    }

    @Provides
    @Singleton
    public CodeTestApplication provideApplication(){
        return this.codeTestApplication;
    }

    @Provides
    @Singleton
    public PreferenceManager providePreferenceManager(CodeTestApplication application) {
        return new PreferenceManager(application);
    }

    @Provides
    @Singleton
    public Moshi provideMoshi() {
        return new Moshi.Builder()
                .add(new Object() {
                    @ToJson
                    public String toJson(UUID uuid) {
                        return uuid.toString();
                    }

                    @FromJson
                    public UUID fromJson(String uuid) {
                        return UUID.fromString(uuid);
                    }
                })
                .add(new Object() {
                    @ToJson
                    public String toJson(TimeZone timeZone) {
                        return timeZone.getID();
                    }

                    @FromJson
                    public TimeZone fromJson(String timezone) {
                        return TimeZone.getTimeZone(timezone);
                    }
                })
                .add(new Object() {
                    @ToJson
                    public Long toJson(Instant instant) {
                        return instant.toEpochMilli();
                    }

                    @FromJson
                    public Instant fromJson(Long instant) {
                        return Instant.ofEpochMilli(instant);
                    }
                })
                .build();
    }

    @Provides
    @Singleton
    public Mc10WebService provideWebService(Moshi moshi) {
        final OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder();

        if (BuildConfig.DEBUG) {
            final HttpLoggingInterceptor logging =
                    new HttpLoggingInterceptor(message -> Timber.d(message)).setLevel(HttpLoggingInterceptor.Level.BODY);

            okHttpClientBuilder.addInterceptor(logging);
        }

        return new Retrofit.Builder()
                .baseUrl("https://qa.mc10cloud.com")
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .client(okHttpClientBuilder.build())
                .build()
                .create(Mc10WebService.class);
    }
}

File: app/src/main/kotlin/com/mc10inc/codetest/di/ApplicationModule.kt
package com.mc10inc.codetest.di

import com.mc10inc.codetest.BuildConfig
import com.mc10inc.codetest.CodeTestApplication
import com.mc10inc.codetest.PreferenceManager
import com.mc10inc.codetest.api.Mc10WebService
import com.squareup.moshi.FromJson
import com.squareup.moshi.Moshi
import com.squareup.moshi.ToJson
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import org.threeten.bp.Instant
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import timber.log.Timber
import java.util.*
import javax.inject.Singleton


/**
 * TODO
 *
 * @author Adam Stroud &#60;[adam.stroud@gmail.com](mailto:adam.stroud@gmail.com)&#62;
 */
@Module
class ApplicationModule(private val codeTestApplication: CodeTestApplication) {
    @Provides
    @Singleton
    fun provideApplication(): CodeTestApplication {
        return this.codeTestApplication
    }

    @Provides
    @Singleton
    fun providePreferenceManager(application: CodeTestApplication): PreferenceManager {
        return PreferenceManager(application)
    }

    @Provides
    @Singleton
    fun provideMoshi(): Moshi {
        return Moshi.Builder()
                .add(object {
                    @ToJson
                    fun toJson(uuid: UUID) = uuid.toString()

                    @FromJson
                    fun fromJson(uuid: String) = UUID.fromString(uuid)
                })
                .add(object {
                    @ToJson
                    fun toJson(timeZone: TimeZone) = timeZone.id

                    @FromJson
                    fun fromJson(timezone: String) = TimeZone.getTimeZone(timezone)
                })
                .add(object {
                    @ToJson
                    fun toJson(instant: Instant) = instant.toEpochMilli()

                    @FromJson
                    fun fromJson(instant: Long) = Instant.ofEpochMilli(instant)
                })
                .build()
    }

    @Provides
    @Singleton
    fun provideWebService(moshi: Moshi): Mc10WebService {
        val okHttpClientBuilder = OkHttpClient.Builder()

        if (BuildConfig.DEBUG) {
            val logging = HttpLoggingInterceptor { message -> Timber.d(message) }
                    .setLevel(HttpLoggingInterceptor.Level.BODY)

            okHttpClientBuilder.addInterceptor(logging)
        }

        return Retrofit.Builder()
                .baseUrl("https://qa.mc10cloud.com")
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .client(okHttpClientBuilder.build())
                .build()
                .create(Mc10WebService::class.java)
    }
}
